kvserver/rangefeed: lift CrdbTestBuild on RangefeedUseBufferedSender

wenyihu6 opened the pull request:
**kvserver/rangefeed: lift CrdbTestBuild on RangefeedUseBufferedSender**

This patch removes the buildutil.CRDBTestBuild check for
RangefeedUseBufferedSender, as the feature is now fully
implemented and safe to use without causing panics.

Epic: none
Release note: none

---

**roachtest: use prng to enable RangefeedUseBufferedSender**

Previously, we used UsingRuntimeAssertions to determine whether
to enable RangefeedUseBufferedSender because it was gated behind
the CRDBTestBuild flag. Now that this check has been removed,
we can use prng to enable the cluster setting 50% of the time.

Note that we still need to avoid enabling the cluster setting
for mixed-version roachtests to prevent it from being applied
to versions that enforce this restriction.

Epic: none
Fixes: #141478
Fixes: #141470
Fixes: #141494
Fixes: #141476
FIxes: #141466
Fixes: #141470
Fixes: #141471
Fixes: #141475
Fixes: #141465

wenyihu6 committed:
kvserver/rangefeed: lift CrdbTestBuild on RangefeedUseBufferedSender

This patch removes the buildutil.CRDBTestBuild check for
RangefeedUseBufferedSender, as the feature is now fully
implemented and safe to use without causing panics.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141595)
<!-- Reviewable:end -->

stevendanna commented on pkg/kv/kvserver/replica_proposal.go:
```diff
@@ -893,6 +893,7 @@ func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult re
 		// blocks waiting for the lease acquisition to finish but it can't finish
 		// because we're not processing raft messages due to holding
 		// processRaftMu (and running on the processRaft goroutine).
+
```
Stray whitespace change.

wenyihu6 replied:
Done.

stevendanna reviewed (approved):
LGTM! Thanks!

wenyihu6 committed:
roachtest: use prng to enable RangefeedUseBufferedSender

Previously, we used UsingRuntimeAssertions to determine whether
to enable RangefeedUseBufferedSender because it was gated behind
the CRDBTestBuild flag. Now that this check has been removed,
we can use prng to enable the cluster setting 50% of the time.

Note that we still need to avoid enabling the cluster setting
for mixed-version roachtests to prevent it from being applied
to versions that enforce this restriction.

Epic: none
Fixes: #141478
Fixes: #141470
Fixes: #141494
Fixes: #141476
FIxes: #141466
Fixes: #141470
Fixes: #141471
Fixes: #141475
Fixes: #141465

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -931,16 +931,9 @@ func (r *testRunner) runWorker(
-				if useBufferedSender {
+				// default.
+				useBufferedSender := prng.Intn(2) == 0
+				if t.IsBuildVersion("v25.2.0") && useBufferedSender {
```
Should we do `t.BuildVersion().AtLeast()` instead so this check doesn't break when we bump the version? Although I'm not sure if either check fixes mixed version tests. Unfortunately the build version referred to here doesn't have any knowledge of the  binaries the mixed version frame work uploads, it's essentially just a check on the version the branch is on.

Maybe the correct fix should be for the mixed version framework to ignore `c.clusterSettings`? Seems like that will always break whenever a new setting is added.

wenyihu6 replied:
Thanks for taking a look! 

> Should we do t.BuildVersion().AtLeast() instead so this check doesn't break when we bump the version?

https://github.com/cockroachdb/cockroach/blob/3ca9845cc23d325eafd8977d5ce76c68b6ace89d/pkg/cmd/roachtest/test_impl.go#L692 

`t.IsBuildVersion` checks for `AtLeast()`. 

> Unfortunately the build version referred to here doesn't have any knowledge of the binaries the mixed version frame work uploads, it's essentially just a check on the version the branch is on.

You are right. Just noticed this from the failing CI 😢. Will try to look into this further and figure out the next steps.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -931,16 +931,9 @@ func (r *testRunner) runWorker(
-				// includes the buildutil.CrdbTestBuild tag. And runtime assertion is
-				// enabled if and only if cockroach binaries used in roachtests include
-				// the buildutil.CrdbTestBuild tag.
-				useBufferedSender := roachtestutil.UsingRuntimeAssertions(t)
```
`UsingRuntimeAssertions` was previously gating benchmark tests from enabling buffered sender. We should probably still disable buffered senders on benchmark tests? Unless you expect there to be no performance impact.

wenyihu6 replied:
We don’t anticipate any negative performance impact, and it would be good to verify if there’s any regression. If we feel strongly about this, the simplest approach might be to opt into this cluster setting when `registry.MetamorphicLeases` is enabled as that registry.TestSpec tries to achieve the same thing 
https://github.com/cockroachdb/cockroach/commit/fc68b0fe8d967ab48d96a659fd5a067787e5e734.

DarrylWong replied:
My concern is less about negative regressions and more about increased variance in benchmark results. i.e. someone working on improving sysbench perf would need to both know about this metamorphic setting and make sure to conditionally enable/disable it or their results will fluctuate more than expected.

That being said I'm not familiar with this work at all so I'll defer to your judgement if you think the perf impact is minimal.

wenyihu6 replied:
I see, that's a good point. I would need to look closer at what the sysbench perf team is analyzing to provide a better answer. I would expect the perf impact to be minimal if the tests they are running don't enable `kv.rangefeed.enabled` but not sure otherwise.

wenyihu6 commented:
TFTR!

bors r=stevendanna, DarrylWong
